src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala

```diff
@@ -1365,16 +1369,18 @@ abstract class BCodeBodyBuilder extends BCodeSkelBuilder {
           argsArray(0) = typeToBType(lambdaTarget.owner.info).toASMType
           i = 1
         }
-        var xs = lambdaTarget.paramss.head
+        var xs = lambdaTargetParamss.head
         while (i < numArgs && (!xs.isEmpty)) {
           argsArray(i) = typeToBType(xs.head.info).toASMType
           i += 1
           xs = xs.tail
         }
         asm.Type.getMethodDescriptor(asmType(functionalInterface), argsArray:_*)
       }
-      val lambdaParams = lambdaTarget.paramss.head.drop(numCaptured)
-      val constrainedType = MethodBType(lambdaParams.map(p => typeToBType(p.tpe)), typeToBType(lambdaTarget.tpe.resultType)).toASMType
+      val lambdaParams = lambdaTargetParamss.head.drop(numCaptured)
+      val lambdaParamsBTypes = new Array[BType](lambdaParams.size)
+      lambdaParams.iterator.map(p => typeToBType(p.tpe)).copyToArray(lambdaParamsBTypes)
+      val constrainedType = MethodBType(lambdaParamsBTypes, typeToBType(lambdaTarget.tpe.resultType)).toASMType
```

Comment:
The `MethodBType` here is just a transient data structure, which we turn immediately into the ASM type and then discard. Instead, could we add to the `MethodBType` object a method that takes the same parameters as the `MethodBType` and directly builds the `ASM type`?